### PR TITLE
Transforms components docs

### PR DESCRIPTION
**Pull request by sfilipi**
Addresses P1 non-Image transforms for #1209. 

Adds samples for Nomalizers, Text, Cocat, Term, KeyToVal

### CODE REVIEW

File: docs/samples/Microsoft.ML.Samples/Dynamic/KeyToValue_Term.cs [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+            // "car truck driver bus pickup horse",  "car truck",                          0
+
+            // A pipeline to convert the terms of the review_reverse column in 
+            // making use of default settings.
+            string defaultColumnName = "DefaultKeys";
+            // REVIEW create through the catalog extension
+            var default_pipeline = new WordTokenizer(ml, "ReviewReverse", "ReviewReverse")
+                .Append(new TermEstimator(ml, "ReviewReverse" , defaultColumnName));
+
+            // Another pipeline, that customizes the advanced settings of the FeaturizeText transformer.
```
Comment (sfilipi) [inline docs/samples/Microsoft.ML.Samples/Dynamic/KeyToValue_Term.cs]: > // Another pip [](start = 11, length = 15)

you guys think I should add more explanations about Reverse?

File: docs/samples/Microsoft.ML.Samples/Dynamic/KeyToValue_Term.cs [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+            {
+                foreach (var value in row.Values)
+                    Console.Write($"{value} ");
+                Console.WriteLine("");
+            }
+
+            // DefaultColumnName column obtained post-transformation.
+            // radiation galaxy universe duck
+            // space galaxy universe radiation
+            // bus pickup universe radiation
```
Comment (sfilipi) [inline docs/samples/Microsoft.ML.Samples/Dynamic/KeyToValue_Term.cs]: >universe radiation [](start = 26, length = 18)

why is this here

File: docs/samples/Microsoft.ML.Samples/Dynamic/KeyToValue_Term.cs [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+            // Create a new ML context, for ML.NET operations. It can be used for exception tracking and logging, 
+            // as well as the source of randomness.
+            var ml = new MLContext(seed: 1, conc: 1);
+
+            // Get a small dataset as an IEnumerable.
+            IEnumerable<SamplesUtils.DatasetUtils.SampleTopicsData> data = SamplesUtils.DatasetUtils.GetTopicsData();
+            var trainData = ml.CreateStreamingDataView(data);
+
+            // Preview of the data.
+            // Review                                    ReviewReverse,                 Label
```
Comment (justinormont) [inline docs/samples/Microsoft.ML.Samples/Dynamic/KeyToValue_Term.cs]:
```suggestion
            // Review,                                    ReviewReverse,                 Label
```

File: docs/samples/Microsoft.ML.Samples/Dynamic/KeyToValue_Term.cs [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+            // "animals birds cats dogs fish horse", "radiation galaxy universe duck",     1
+            // "horse birds house fish duck cats",   "space galaxy universe radiation",    0
+            // "car truck driver bus pickup",        "bus pickup",                         1
+            // "car truck driver bus pickup horse",  "car truck",                          0
```
Comment (justinormont) [inline docs/samples/Microsoft.ML.Samples/Dynamic/KeyToValue_Term.cs]: May want to say the goal of the dataset. Eg: _"The goal of the dataset to classify if the review matches ..."_

I ask this, mainly as I'm reading the example, I have no idea what the labels represent vs. the data.

File: docs/samples/Microsoft.ML.Samples/Dynamic/KeyToValue_Term.cs [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+            // A pipeline to convert the terms of the review_reverse column in 
+            // making use of default settings.
+            string defaultColumnName = "DefaultKeys";
+            // REVIEW create through the catalog extension
+            var default_pipeline = new WordTokenizer(ml, "ReviewReverse", "ReviewReverse")
+                .Append(new TermEstimator(ml, "ReviewReverse" , defaultColumnName));
+
+            // Another pipeline, that customizes the advanced settings of the TermEstimator.
```
Comment (justinormont) [inline docs/samples/Microsoft.ML.Samples/Dynamic/KeyToValue_Term.cs]: May want to say _why_ changing the hyperparameters of the TermEstimator is useful. Why keep only the first 10 ASCIIbetically ordered terms?

File: docs/samples/Microsoft.ML.Samples/Dynamic/KeyToValue_Term.cs [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+            // "animals birds cats dogs fish horse", "radiation galaxy universe duck",     1
+            // "horse birds house fish duck cats",   "space galaxy universe radiation",    0
+            // "car truck driver bus pickup",        "bus pickup",                         1
+            // "car truck driver bus pickup horse",  "car truck",                          0
+
+            // A pipeline to convert the terms of the review_reverse column in 
+            // making use of default settings.
+            string defaultColumnName = "DefaultKeys";
+            // REVIEW create through the catalog extension
+            var default_pipeline = new WordTokenizer(ml, "ReviewReverse", "ReviewReverse")
```
Comment (justinormont) [inline docs/samples/Microsoft.ML.Samples/Dynamic/KeyToValue_Term.cs]: Why use WordTokenizer+Term instead of TextTransform?

**Comment by justinormont** [comment]
The two added datasets could use an entry in the dataset README.md: https://github.com/dotnet/machinelearning/blob/586533cb80d7e0ea18c02f2e510e1a5cc0e27f41/test/data/README.md

File: docs/samples/Microsoft.ML.Samples/Dynamic/KeyToValue_Term.cs (hunk 6 above)
Comment (sfilipi) [reply docs/samples/Microsoft.ML.Samples/Dynamic/KeyToValue_Term.cs]: The sample is about TermEstimator. 

---
In reply to: [227237000](https://github.com/dotnet/machinelearning/pull/1321#discussion_r227237000) [](ancestors = 227237000)

File: docs/samples/Microsoft.ML.Samples/Dynamic/KeyToValue_Term.cs (hunk 5 above)
Comment (sfilipi) [reply docs/samples/Microsoft.ML.Samples/Dynamic/KeyToValue_Term.cs]: see if you'd phrase it differently. Maybe we can leave longer explanations for the documentation?

---
In reply to: [227236905](https://github.com/dotnet/machinelearning/pull/1321#discussion_r227236905) [](ancestors = 227236905)

File: docs/samples/Microsoft.ML.Samples/Dynamic/ConcatTransform.cs [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+            // A pipeline for concatenating the age, parity and induced columns together in the Features column.
+            string outputColumnName = "Features";
+            var pipeline = new ConcatEstimator(ml, outputColumnName, new[] { "Age", "Parity", "Induced"});
+
+            // The transformed data.
+            var transformedData = pipeline.Fit(trainData).Transform(trainData);
+
+            // Getting the data of the newly created column as an IEnumerable of SampleInfertDataWithFeatures.
+            var featuresColumn = transformedData.AsEnumerable<SampleInfertDataWithFeatures>(ml, reuseRowObject: false);
```
Comment (GalOshri) [inline docs/samples/Microsoft.ML.Samples/Dynamic/ConcatTransform.cs]: What does `reuseRowObject` do? Would defaults be fine for these samples?

Comment (sfilipi) [reply docs/samples/Microsoft.ML.Samples/Dynamic/ConcatTransform.cs]: for the enumerable, it determines whether to return the same object on every row, or allocate a new one per row. It is a required param; doesn't have a default. 

For the settings of the transforms, i am using both defaults and non-defaults; since the purpose of this snippet is to educate about usage. 

---
In reply to: [227591855](https://github.com/dotnet/machinelearning/pull/1321#discussion_r227591855) [](ancestors = 227591855)

File: docs/samples/Microsoft.ML.Samples/Dynamic/MinMaxNormalizer.cs [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+            {
+                Console.WriteLine($"{colName} column obtained post-transformation.");
+                foreach (var row in column)
+                    Console.WriteLine($"{row} ");
+            };
+
+            printHelper("Induced", normalizedColumn);
+
+            // Preview of the data.
+            // Induced
```
Comment (shmoradims) [inline docs/samples/Microsoft.ML.Samples/Dynamic/MinMaxNormalizer.cs]: >Induced [](start = 15, length = 7)

nit: this doesn't match $"{colName} column obtained post-transformation."

File: docs/samples/Microsoft.ML.Samples/Static/FastTree.cs [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+
+            // Creating a data reader, based on the format of the data
+            var reader = TextLoader.CreateReader(env, c => (
+                        label: c.LoadFloat(0),
+                        features: c.LoadFloat(1, 6)
+                    ),
+                separator: '\t', hasHeader: true);
+
+            // Read the data, and leave 10% out, so we can use them for testing
+            var data = reader.Read(new MultiFileSource(dataFile));
```
Comment (shmoradims) [inline docs/samples/Microsoft.ML.Samples/Static/FastTree.cs]: >new MultiFileSource(dataFile) [](start = 35, length = 29)

I think you can use dataFile directly for a single file.

File: docs/samples/Microsoft.ML.Samples/Static/FastTree.cs [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+                L2: cvResults.Select(r => r.metrics.L2).Average(),
+                LossFn: cvResults.Select(r => r.metrics.LossFn).Average(),
+                Rms: cvResults.Select(r => r.metrics.Rms).Average(),
+                RSquared: cvResults.Select(r => r.metrics.RSquared).Average()
+            );
+            Console.WriteLine($"L1 - {averagedMetrics.L1}");
+            Console.WriteLine($"L2 - {averagedMetrics.L2}");
+            Console.WriteLine($"LossFunction - {averagedMetrics.LossFn}");
+            Console.WriteLine($"RMS - {averagedMetrics.Rms}");
+            Console.WriteLine($"RSquared - {averagedMetrics.RSquared}");
```
Comment (shmoradims) [inline docs/samples/Microsoft.ML.Samples/Static/FastTree.cs]: please add preview

File: docs/samples/Microsoft.ML.Samples/Dynamic/MinMaxNormalizer.cs (hunk 8 above)
Comment (sfilipi) [reply docs/samples/Microsoft.ML.Samples/Dynamic/MinMaxNormalizer.cs]: The col name is "Induced", see line 39. 

---
In reply to: [227974775](https://github.com/dotnet/machinelearning/pull/1321#discussion_r227974775) [](ancestors = 227974775)

File: docs/samples/Microsoft.ML.Samples/Dynamic/KeyToValue_Term.cs [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+            // Create a new ML context, for ML.NET operations. It can be used for exception tracking and logging, 
+            // as well as the source of randomness.
+            var ml = new MLContext(seed: 1, conc: 1);
+
+            // Get a small dataset as an IEnumerable.
+            IEnumerable<SamplesUtils.DatasetUtils.SampleTopicsData> data = SamplesUtils.DatasetUtils.GetTopicsData();
+            var trainData = ml.CreateStreamingDataView(data);
+
+            // Preview of the topics data; a dataset that contains one column with two set of keys assigned to a body of text
+            // Review and ReviewReverse. The dataset will be used to classify how accurately the keys are assigned to the text.
```
Comment (shmoradims) [inline docs/samples/Microsoft.ML.Samples/Dynamic/KeyToValue_Term.cs]: >ReviewReverse [](start = 26, length = 13)

It's not clear to me what this column means.

File: docs/samples/Microsoft.ML.Samples/Dynamic/KeyToValue_Term.cs (hunk 4 above)
Comment (shmoradims) [reply docs/samples/Microsoft.ML.Samples/Dynamic/KeyToValue_Term.cs]: +1. It's not obvious what the labels mean.

---
In reply to: [227236398](https://github.com/dotnet/machinelearning/pull/1321#discussion_r227236398) [](ancestors = 227236398)

File: docs/samples/Microsoft.ML.Samples/Dynamic/KeyToValue_Term.cs [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+            // "animals birds cats dogs fish horse", "radiation galaxy universe duck",     1
+            // "horse birds house fish duck cats",   "space galaxy universe radiation",    0
+            // "car truck driver bus pickup",        "bus pickup",                         1
+            // "car truck driver bus pickup horse",  "car truck",                          0
+
+            // A pipeline to convert the terms of the review_reverse column in 
+            // making use of default settings.
+            string defaultColumnName = "DefaultKeys";
+            // REVIEW create through the catalog extension
+            var default_pipeline = new WordTokenizeEstimator(ml, "ReviewReverse", "ReviewReverse")
```
Comment (shmoradims) [inline docs/samples/Microsoft.ML.Samples/Dynamic/KeyToValue_Term.cs]: >ReviewReverse [](start = 83, length = 13)

I think it's better to leave this out, if you want to do the transformation in-place.

File: docs/samples/Microsoft.ML.Samples/Dynamic/KeyToValue_Term.cs [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+                foreach (var value in row.Values)
+                    Console.Write($"{value} ");
+                Console.WriteLine("");
+            }
+
+            // DefaultKeys 
+            // radiation galaxy universe duck
+            // space galaxy universe radiation
+            // bus pickup universe radiation
+            // car truck universe radiation
```
Comment (shmoradims) [inline docs/samples/Microsoft.ML.Samples/Dynamic/KeyToValue_Term.cs]: >universe radiation [](start = 25, length = 18)

this seems to be a bug too.

File: docs/samples/Microsoft.ML.Samples/Dynamic/MinMaxNormalizer.cs (hunk 8 above)
Comment (shmoradims) [reply docs/samples/Microsoft.ML.Samples/Dynamic/MinMaxNormalizer.cs]: I meant if these comment are supposed to be the output from printHelper. If they are, then the headers don't match:
            // Preview of the data.
            // Induced

vs 

Console.WriteLine($"{colName} column obtained post-transformation.");


---
In reply to: [227982369](https://github.com/dotnet/machinelearning/pull/1321#discussion_r227982369) [](ancestors = 227982369,227974775)

File: docs/samples/Microsoft.ML.Samples/Dynamic/KeyToValue_Term.cs (hunk 13 above)
Comment (sfilipi) [reply docs/samples/Microsoft.ML.Samples/Dynamic/KeyToValue_Term.cs]: right, will investigate post PR. 

---
In reply to: [228249405](https://github.com/dotnet/machinelearning/pull/1321#discussion_r228249405) [](ancestors = 228249405)

File: docs/samples/Microsoft.ML.Samples/Dynamic/KeyToValue_Term.cs [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+            // REVIEW create through the catalog extension
+            var default_pipeline = new WordTokenizeEstimator(ml, "ReviewReverse")
+                .Append(new TermEstimator(ml, "ReviewReverse" , defaultColumnName));
+
+            // Another pipeline, that customizes the advanced settings of the TermEstimator.
+            // We can change the maxNumTerm to limit how many keys will get generated out of the set of words, 
+            // and condition the order in which they get evaluated by changing sort from the default Occurence (order in which they get encountered) 
+            // to value/alphabetically.
+            string customizedColumnName = "CustomizedKeys";
+            var customized_pipeline = new WordTokenizeEstimator(ml, "ReviewReverse", "ReviewReverse")
```
Comment (shmoradims) [inline docs/samples/Microsoft.ML.Samples/Dynamic/KeyToValue_Term.cs]: >ReviewReverse [](start = 86, length = 13)

plz remove for in-place transformation like above

File: docs/samples/Microsoft.ML.Samples/Program.cs [hunk 15]
```diff
@@ -2,14 +2,16 @@
...
 namespace Microsoft.ML.Samples
 {
     internal static class Program
     {
         static void Main(string[] args)
         {
-            Trainers.SdcaRegression();
-            Transformers.ConcatEstimator();
+            TransformSamples.MinMaxNormalizer();
+            // TrainersSamples.LightGbmRegression();
```
Comment (Ivanidzo4ka) [inline docs/samples/Microsoft.ML.Samples/Program.cs]: >// TrainersSamples.LightGbmRegression(); [](start = 12, length = 40)

looks weird.

File: docs/samples/Microsoft.ML.Samples/Static/FastTree.cs [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+    {
+        public static void FastTreeRegression()
+        {
+            // Downloading a regression dataset from github.com/dotnet/machinelearning
+            // this will create a housing.txt file in the filsystem this code will run
+            // you can open the file to see the data. 
+            string dataFile = SamplesUtils.DatasetUtils.DownloadHousingRegressionDataset();
+
+            // Creating the ML.Net IHostEnvironment object, needed for the pipeline
+            var env = new LocalEnvironment(seed: 0);
```
Comment (Ivanidzo4ka) [inline docs/samples/Microsoft.ML.Samples/Static/FastTree.cs]: > var env = new LocalEnvironment(seed: 0); [](start = 11, length = 41)

why for transforms examples we use mlContext, and here we create LocalEnvironment?

File: test/data/gplv2/airquality.csv [hunk 17]
```diff
@@ -0,0 +1,154 @@
...
+"3",12,149,12.6,74,5,3
+"4",18,313,11.5,62,5,4
+"5",NA,NA,14.3,56,5,5
+"6",28,NA,14.9,66,5,6
+"7",23,299,8.6,65,5,7
+"8",19,99,13.8,59,5,8
+"9",8,19,20.1,61,5,9
+"10",NA,194,8.6,69,5,10
+"11",7,NA,6.9,74,5,11
+"12",16,256,9.7,69,5,12
```
Comment (Ivanidzo4ka) [inline test/data/gplv2/airquality.csv]: I don't see any usage of this file in PR, not even mentions of air quality.
Is it necessary?
And is it regression dataset?  because we have [https://github.com/dotnet/machinelearning/issues/938](https://github.com/dotnet/machinelearning/issues/938)

Comment (sfilipi) [reply test/data/gplv2/airquality.csv]: planning to keep the samples consistent with our other repo. They use this dataset, so it will get used in the next set of samples. Just added the two we need in one shot .

---
In reply to: [228309756](https://github.com/dotnet/machinelearning/pull/1321#discussion_r228309756) [](ancestors = 228309756)

File: docs/samples/Microsoft.ML.Samples/Static/FastTree.cs (hunk 16 above)
Comment (sfilipi) [reply docs/samples/Microsoft.ML.Samples/Static/FastTree.cs]: thanks for spotting it. I updated them. 

---
In reply to: [228308134](https://github.com/dotnet/machinelearning/pull/1321#discussion_r228308134) [](ancestors = 228308134)

File: docs/samples/Microsoft.ML.Samples/Dynamic/KeyToValue_Term.cs [hunk 18]
```diff
@@ -0,0 +1,115 @@
...
+        using Microsoft.ML.Transforms.Text;
+        using System;
+        using System.Collections.Generic;
+        using System.Linq;
+
+namespace Microsoft.ML.Samples.Dynamic
+{
+    public partial class TransformSamples
+    {
+        public static void KeyToValue_Term()
```
Comment (Ivanidzo4ka) [inline docs/samples/Microsoft.ML.Samples/Dynamic/KeyToValue_Term.cs]: >KeyToValue_Term [](start = 27, length = 15)

This is standing out, what this "_" mean, and why it cannot be KeyToValueAndTerm or KeyToValueThenTerm?
